flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/allocator/DefaultSlotAssigner.java

```diff
@@ -43,35 +49,86 @@ public Collection<SlotAssignment> assignSlots(
         }
 
-        Iterator<? extends SlotInfo> iterator = freeSlots.iterator();
+        Collection<? extends SlotInfo> pickedSlots = freeSlots;
+        // To avoid the sort-work loading.
+        if (freeSlots.size() > allExecutionSlotSharingGroups.size()) {
+            final Map<TaskManagerLocation, ? extends Set<? extends SlotInfo>> slotsPerTaskExecutor =
+                    getSlotsPerTaskExecutor(freeSlots);
+            pickedSlots =
+                    pickSlotsInMinimalTaskExecutors(
+                            slotsPerTaskExecutor,
+                            allExecutionSlotSharingGroups.size(),
+                            getSortedTaskExecutors(slotsPerTaskExecutor));
+        }
+
+        Iterator<? extends SlotInfo> iterator = pickedSlots.iterator();
         Collection<SlotAssignment> assignments = new ArrayList<>();
-        for (ExecutionSlotSharingGroup group : allGroups) {
+        for (ExecutionSlotSharingGroup group : allExecutionSlotSharingGroups) {
             assignments.add(new SlotAssignment(iterator.next(), group));
         }
         return assignments;
     }
 
-    static List<ExecutionSlotSharingGroup> createExecutionSlotSharingGroups(
-            VertexParallelism vertexParallelism, SlotSharingGroup slotSharingGroup) {
-        final Map<Integer, Set<ExecutionVertexID>> sharedSlotToVertexAssignment = new HashMap<>();
-        slotSharingGroup
-                .getJobVertexIds()
-                .forEach(
-                        jobVertexId -> {
-                            int parallelism = vertexParallelism.getParallelism(jobVertexId);
-                            for (int subtaskIdx = 0; subtaskIdx < parallelism; subtaskIdx++) {
-                                sharedSlotToVertexAssignment
-                                        .computeIfAbsent(subtaskIdx, ignored -> new HashSet<>())
-                                        .add(new ExecutionVertexID(jobVertexId, subtaskIdx));
-                            }
-                        });
-        return sharedSlotToVertexAssignment.values().stream()
-                .map(ExecutionSlotSharingGroup::new)
-                .collect(Collectors.toList());
+    /**
+     * In order to minimize the using of task executors at the resource manager side in the
+     * session-mode and release more task executors in a timely manner, it is a good choice to
+     * prioritize selecting slots on task executors with the least available slots. This strategy
+     * also ensures that relatively fewer task executors can be used in application-mode.
+     */
+    private Iterator<TaskManagerLocation> getSortedTaskExecutors(
+            Map<TaskManagerLocation, ? extends Set<? extends SlotInfo>> slotsPerTaskExecutor) {
+        final Comparator<TaskManagerLocation> taskExecutorComparator =
+                Comparator.comparingInt(tml -> slotsPerTaskExecutor.get(tml).size());
+        return sortTaskExecutors(slotsPerTaskExecutor.keySet(), taskExecutorComparator);
```

Comment:
It seems `sortTaskExecutors` method is not needed, and we could call ` taskManagerLocations.stream().sorted(taskExecutorComparator).iterator()` directly.